security/pkg/k8s/chiron/controller.go

```diff
@@ -0,0 +1,530 @@
+
+	svcName, found := wc.getServiceName(scrtName)
+	if !found {
+		return fmt.Errorf("failed to find the service name for the secret (%v) to refresh", scrtName)
+	}
+
+	chain, key, err := GenKeyCertK8sCA(wc, scrtName, namespace, svcName)
+	if err != nil {
+		return err
+	}
+
+	scrt.Data[CertChainID] = chain
+	scrt.Data[PrivateKeyID] = key
+	scrt.Data[RootCertID] = wc.getCurCACert()
+
+	_, err = wc.core.Secrets(namespace).Update(scrt)
+	return err
+}
+
+// Clean up the CSR
+func (wc *WebhookController) cleanUpCertGen(csrName string) error {
+	// Delete CSR
+	log.Debugf("delete CSR: %v", csrName)
+	err := wc.certClient.CertificateSigningRequests().Delete(csrName, nil)
+	if err != nil {
+		log.Errorf("failed to delete CSR (%v): %v", csrName, err)
+		return err
+	}
+	return nil
+}
+
+// Return whether the input secret name is a Webhook secret
+func (wc *WebhookController) isWebhookSecret(name, namespace string) bool {
+	for _, n := range WebhookSecretNames {
+		if name == n && namespace == wc.namespace {
+			return true
+		}
+	}
+	return false
+}
+
+func (wc *WebhookController) watchCACert(stopCh chan struct{}) {
+	var timerC <-chan time.Time
+	for {
+		select {
+		case <-timerC:
+			// Update webhook certificates and WebhookConfigurations
+			timerC = nil
+
+			caCert, err := readCACert()
+			if err != nil {
+				log.Errorf("failed to read CA certificate: %v", err)
+				break
+			}
+			if !bytes.Equal(caCert, wc.getCurCACert()) {
+				log.Debug("CA cert changed, update webhook certs and webhook configuration")
+				wc.setCurCACert(caCert)
+				// Update the webhook certificates
+				for _, name := range WebhookSecretNames {
+					wc.upsertSecret(name, wc.namespace)
+				}
+				// Patch the WebhookConfiguration
+				// TODO (lei-tang): extend the demo webhook to all webhooks
+				doPatch(wc.k8sClient, wc.mutatingWebhookConfigNames[0], wc.mutatingWebhookNames[0], wc.getCurCACert())
+			}
+		case event := <-wc.CaCertWatcher.Event:
+			// use a timer to debounce configuration updates
+			if (event.IsModify() || event.IsCreate()) && timerC == nil {
+				timerC = time.After(watchDebounceDelay)
+			}
+		case err := <-wc.CaCertWatcher.Error:
+			log.Errorf("CA certificate watcher error: %v", err)
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (wc *WebhookController) getCurCACert() []byte {
+	wc.mutex.Lock()
+	cp := append([]byte(nil), wc.curCACert...)
+	wc.mutex.Unlock()
+	return cp
+}
+
+func (wc *WebhookController) setCurCACert(cert []byte) {
+	wc.mutex.Lock()
+	wc.curCACert = append([]byte(nil), cert...)
+	wc.mutex.Unlock()
+}
+
+// Get the service name for the secret. Return the service name and whether it is found.
+func (wc *WebhookController) getServiceName(secretName string) (string, bool) {
+	for i, name := range WebhookSecretNames {
+		if name == secretName {
+			return WebhookServiceNames[i], true
+		}
+	}
+	return "", false
+}
+
+func (wc *WebhookController) checkAndPatchMutatingWebhook(host string, port int, stopCh chan struct{}) {
```

Comment:
Thanks for adding this to the prototype. This is one of the parts I've worried about most. 

The current code delays _patching_ the webhook until the endpoint is ready. However, the webhook is still created with an empty caBundle. The apiserver would try to use this webhook but fail because of "unknown authority". I think this needs to be tweaked such that the webhook controller also _creates_ the webhook configuration, which could be conditional on the corresponding endpoint being ready. That would providing the webhook template(s) to the controller, e.g. via configmap or something about API.

Reply:
Done: changed according to the suggestion. Thanks!